README.md
# Vintage Heart

For [Nerves Project](https://nerves-project.org) apps only, specifically those using [VintageNet](https://hexdocs.pm/vintage_net/VintageNet.html) over WiFi. Really only used with Raspberry Pi Zero W's connecting over WiFi on "wlan0".

Crudely solves a problem that I have intermittenly encountered with connectivity being lost, while the IP address remains assigned. Being intermittent it is a devil to debug so this is what I use instead.


It does this by doing the following:

* Registering a the [`:heart` status callback](https://github.com/nerves-project/nerves_heart).
* Checking the `["interface", "wlan0", "connection"]` [VintageNet property](https://hexdocs.pm/vintage_net/readme.html#properties) every 10 seconds (by default). If `:internet` is returned then we assume everything is good (and counters are reset). Otherwise counters are incremented.
* On being offline for about 4 minutes (by default), VintageNet is given a bit of a kick. A kick being killing the `VintageNet.RouteManager` GenServer. This is a bit brutal, but usually does the job of going out and reconnecting with the network. 
* If we're offline for 14 minutes (by default, which will by default involve 3 kicks) we'll report to `:heart` that we are no longer ok, and it will reboot the device.

Note that 

* If the "wlan0" IP address is _192.168.0.1_ (or as otherwise configured) then it's assumed that the [VintageNetWizard](https://hexdocs.pm/vintage_net_wizard/readme.html) is being used to set up the WiFi connection, and checks will not happen
* If you compile with the default `MIX_TARGET` being host, then checking is also essentially disabled. That is it will not get in the way of you running `iex -S mix` to try out your "nerves" code on your development machine. 
* Connectivity is determined by having an "Internet connection", rather than a network connection. You can configure what having an "Internet connection" means via [VintageNet](https://hexdocs.pm/vintage_net/readme.html#internet-connectivity-checks)

This is a straight extraction, more or less, from an existing project.

## Has anything happened?

You might be interested to know if anything has happened. You can check with

```elixir

iex(1)> VintageHeart.Pulse.full_status()
%VintageHeart.Pulse{
  offline_this_period_count: 0,
  offline_count: 0,
  status: :ok,
  highest_offline: 25,
  last_kick: ~U[2023-01-19 15:35:55.603980Z]
}
```

Reboots aren't logged though you can check uptime for that.


## Configuration

If you do not like the defaults then you can add some configuration to your project. The following example does nothing, in that it pointlessly confirms the default values, but you might want to change them.


```elixir
config :vintage_heart, VintageHeart.Pulse
  heart_callback?: true, #  Whether to use the status as `:heart` callback
  poll_interval_millis: :timer.seconds(10), # how often to poll
  offline_kick_count: 24, # how many polls without an internet connection before giving VintageNet a kick
  offline_status_down_count: 84, # how many polls without an internet connection before setting the status to down
  wizard_hotspot_ip: {192, 168, 0, 1} # IP used if VintageNetWizard is active as a hotspot
```

## Using

[Now an hexicle](https://hex.pm/packages/vintage_heart)

```elixir

defp deps do
  [
    # etc...
     {:vintage_heart, "~> 0.1"},
     # etc ... 
  ]
end
```

CHANGELOG.md
# Change log

# v0.1.1
* Introducded changelog
* Fixed GenServer call matching bug with `VintageHeart.Pulse.last_kick/0`
* Added `VintageHeart.Pulse.full_status/0`

# v0.1.2 

* Fixed a doc error

test/vintage_heart/pulse_test.exs
# credo:disable-for-this-file Credo.Check.Readability.LargeNumbers
defmodule VintageHeart.PulseTest do
  use ExUnit.Case

  alias VintageHeart.Pulse

  @tag capture_log: true
  test "initialised with empty counts" do
    assert {:ok, %{status: :ok, offline_count: 0, offline_this_period_count: 0}} = Pulse.init(:_)
  end

  test "resets counters when the VintagetNetWizard hotspot is up" do
    set_ip4({192, 168, 0, 1})
    stub_prop(:lan)

    assert {:noreply,
            %{status: :ok, offline_count: 0, offline_this_period_count: 0, highest_offline: 4}} =
             Pulse.handle_info(:check, %Pulse{
               offline_count: 2,
               offline_this_period_count: 2,
               highest_offline: 4
             })
  end

  test "offline and lan_only incremented when no ip" do
    stub_prop(["interface", "wlan0", "addresses"], [])
    stub_prop(:disconnected)

    assert {:noreply, %{status: :ok, offline_count: 5, offline_this_period_count: 3}} =
             Pulse.handle_info(:check, %Pulse{offline_count: 4, offline_this_period_count: 2})
  end

  test "resets counters when connected" do
    stub_prop(:internet)

    assert {:noreply,
            %{status: :ok, offline_count: 0, offline_this_period_count: 0, highest_offline: 3}} =
             Pulse.handle_info(:check, %Pulse{
               offline_count: 2,
               offline_this_period_count: 2,
               highest_offline: 3
             })
  end

  test "increments counters when lan, and ip4 is not the hotspot and not on internet" do
    set_ip4({192, 168, 1, 12})
    stub_prop(:lan)

    assert {:noreply, %{status: :ok, offline_count: 5, offline_this_period_count: 3}} =
             Pulse.handle_info(:check, %Pulse{offline_count: 4, offline_this_period_count: 2})
  end

  test "resets lan count, increments offline_count, and kicks VintageNet on 24th lan_only" do
    stub_prop(:lan)

    assert {:noreply,
            %{
              status: :ok,
              offline_count: 41,
              offline_this_period_count: 0,
              last_kick: %DateTime{}
            }} =
             Pulse.handle_info(:check, %Pulse{offline_count: 40, offline_this_period_count: 23})

    assert_receive :vintage_net_kicked
  end

  test "highest_offline is the highest achieved" do
    assert {:noreply, %{highest_offline: 4, offline_count: 3}} =
             Pulse.handle_info(:check, %Pulse{offline_count: 2, highest_offline: 4})

    assert {:noreply, %{highest_offline: 5}} =
             Pulse.handle_info(:check, %Pulse{offline_count: 4, highest_offline: 4})
  end

  test "is not ok after 84 offlines" do
    assert {:noreply, %{status: :down}} =
             Pulse.handle_info(:check, %Pulse{offline_count: 83, offline_this_period_count: 2})
  end

  test "status not reset at 85 n_nets" do
    assert {:noreply, %{status: :down}} =
             Pulse.handle_info(:check, %Pulse{
               offline_count: 85,
               offline_this_period_count: 2,
               status: :down
             })
  end

  test "smoke test the GenServer calls" do
    # Make sure these simple methods are properly wired up, which
    # has not always been the case

    assert nil == Pulse.last_kick()
    assert :ok == Pulse.status()
    assert %Pulse{} = Pulse.full_status()
  end

  defp set_ip4(addr) do
    addresses = [
      %{
        address: addr,
        family: :inet,
        netmask: {255, 255, 255, 0},
        prefix_length: 24,
        scope: :universe
      },
      %{
        address: {65152, 0, 0, 0, 47655, 60415, 65071, 30543},
        family: :inet6,
        netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0},
        prefix_length: 64,
        scope: :link
      }
    ]

    stub_prop(["interface", "wlan0", "addresses"], addresses)
  end

  defp stub_prop(property \\ ["interface", "wlan0", "connection"], value) do
    send(self(), {:vn_prop, property, value})
  end
end


test/support/testing_vintagenet.ex
defmodule VintageHeart.TestingVintagenet do
  @moduledoc false
  @behaviour VintageHeart.Vintagenet
  @doc """
  If the current process has a message with
  `{:vn_prop, property, value}` then value is returned
  else nil
  """
  @impl true
  def get_properties(property) do
    receive do
      {:vn_prop, ^property, value} -> value
    after
      1 -> nil
    end
  end

  @doc """
  Sends :vintage_net_kicked to the current process
  """
  @impl true
  def kick do
    send(self(), :vintage_net_kicked)
  end
end


test/vintage_heart/vintagenet_test.exs
defmodule VintageHeart.VintagenetTest do
  use ExUnit.Case
  alias VintageHeart.Vintagenet

  test "test always resolves to `VintageHeart.TestingVintagenet`" do
    assert VintageHeart.TestingVintagenet == Vintagenet.implementation(:test, :host)
    assert VintageHeart.TestingVintagenet == Vintagenet.implementation(:test, :other)
    assert VintageHeart.TestingVintagenet == Vintagenet.implementation(:test, :rpi0)
  end

  test "when env is not test, then when target is host then the implementation is `VintageHeart.StubVintagenet" do
    assert VintageHeart.StubVintagenet == Vintagenet.implementation(:dev, :host)
    assert VintageHeart.StubVintagenet == Vintagenet.implementation(:prod, :host)
  end

  test "when env is not test, and the target is not host, then the implementation is the real deal" do
    assert VintageHeart.RealVintagenet == Vintagenet.implementation(:dev, :rpi0)
    assert VintageHeart.RealVintagenet == Vintagenet.implementation(:prod, :rpi0)
    assert VintageHeart.RealVintagenet == Vintagenet.implementation(:prod, :rpi3)
  end
end


mix.exs
defmodule VintageHeart.MixProject do
  use Mix.Project

  def project do
    [
      app: :vintage_heart,
      version: "0.1.2",
      elixir: "~> 1.13",
      start_permanent: Mix.env() == :prod,
      elixirc_paths: elixirc_paths(Mix.env()),
      deps: deps(),
      package: package(),
      docs: docs()
    ]
  end

  def application do
    [
      extra_applications: [:logger],
      mod: {VintageHeart.Application, []}
    ]
  end

  defp deps do
    [
      {:credo, "~> 1.6", only: [:dev, :test]},
      {:dialyxir, "~> 1.2", only: [:dev, :test], runtime: false},
      {:ex_doc, "~> 0.29.1", only: :dev, runtime: false}
    ]
  end

  defp elixirc_paths(:test), do: ["lib", "test/support"]
  defp elixirc_paths(_), do: ["lib"]

  defp package() do
    [
      description:
        "For Nerves projects that need an Internet connection, monitors and gives VintageNet a kick and eventually causes a reboot if connectivity is lost",
      licenses: ["Apache-2.0"],
      links: %{"GitHub" => "https://github.com/paulanthonywilson/vintage_heart/"}
    ]
  end

  defp docs do
    [main: "readme", extras: ["README.md", "CHANGELOG.md"]]
  end
end


lib/vintage_heart/pulse.ex
defmodule VintageHeart.Pulse do
  @moduledoc """
  VintageHeart.

  * Without an internet connection for about 14 minutes with a 30 second check interval) reboots
  * After about 4 mins with only a 'lan' connection (no internet) kicks VintageNet by killing
  `VintageNet.RouteManager`

  All check counts are reset to zero on every heartbeat if the local ip address is 192.168.0.1, ie
  we are running the hotspot for setting up the internet connection.
  """
  use GenServer
  use VintageHeart.Vintagenet

  alias VintageHeart.Configuration

  @name __MODULE__
  require Logger

  defstruct offline_this_period_count: 0,
            offline_count: 0,
            status: :ok,
            highest_offline: 0,
            last_kick: nil

  @type t :: %__MODULE__{
          offline_this_period_count: pos_integer(),
          offline_count: pos_integer(),
          status: :ok | :down,
          highest_offline: pos_integer(),
          last_kick: nil | DateTime.t()
        }
  def(start_link(opts)) do
    GenServer.start_link(__MODULE__, opts, name: @name)
  end

  def status do
    GenServer.call(@name, :status)
  end

  @doc """
  When was the last time the newtwork kicked?
  """
  @spec last_kick :: nil | DateTime.t()
  def last_kick do
    GenServer.call(@name, :last_kick)
  end

  @doc """
  All the status
  """
  @spec full_status :: t()
  def full_status do
    GenServer.call(@name, :full_status)
  end

  def init(_) do
    if Process.whereis(:heart) do
      :heart.set_callback(__MODULE__, :status)
    else
      Logger.info("No heartbeat set")
    end

    schedule_poll()

    {:ok, %__MODULE__{}}
  end

  def handle_call(:status, _, %{status: status} = s) do
    {:reply, status, s}
  end

  def handle_call(:last_kick, _, %{last_kick: last_kick} = s) do
    {:reply, last_kick, s}
  end

  def handle_call(:full_status, _, s) do
    {:reply, s, s}
  end

  def handle_info(:check, s) do
    schedule_poll()
    {:noreply, check(s)}
  end

  defp check(state) do
    if hotspot?() do
      reset_counters(state)
    else
      state
      |> check_connection()
      |> maybe_kick_vintage_net()
      |> maybe_set_status_to_down()
    end
  end

  defp check_connection(
         %{
           offline_this_period_count: lan_count,
           offline_count: net_count,
           highest_offline: highest_offline
         } = s
       ) do
    if :internet == connection_status() do
      reset_counters(s)
    else
      offline_count = net_count + 1

      %{
        s
        | offline_this_period_count: lan_count + 1,
          offline_count: offline_count,
          highest_offline: Enum.max([offline_count, highest_offline])
      }
    end
  end

  defp reset_counters(state) do
    %{state | offline_this_period_count: 0, offline_count: 0}
  end

  defp connection_status do
    Vintagenet.get_properties(["interface", "wlan0", "connection"])
  end

  defp maybe_kick_vintage_net(%{offline_this_period_count: offline_this_period_count} = s) do
    if offline_this_period_count == Configuration.offline_kick_count() do
      Vintagenet.kick()
      %{s | offline_this_period_count: 0, last_kick: DateTime.utc_now()}
    else
      s
    end
  end

  defp maybe_set_status_to_down(%{offline_count: offline_count} = s) do
    if offline_count == Configuration.offline_status_down_count() do
      %{s | status: :down}
    else
      s
    end
  end

  defp hotspot?, do: Configuration.wizard_hotspot_ip() == ipv4()

  defp ipv4 do
    ["interface", "wlan0", "addresses"]
    |> Vintagenet.get_properties()
    |> find_ipv4()
  end

  defp schedule_poll do
    Process.send_after(self(), :check, Configuration.poll_interval_millis())
  end

  defp find_ipv4(nil), do: nil

  defp find_ipv4(addresses) do
    addresses
    |> Enum.find(%{}, fn %{family: family} -> family == :inet end)
    |> Map.get(:address)
  end
end


lib/vintage_heart/vintagenet.ex
defmodule VintageHeart.Vintagenet do
  @moduledoc """
  Indirection layer for `VintageNet` used within this project so we can test etc... off
  the hardware. Usage

  ```
  use VintageHeart.Vintagenet
  ```

  The appropriate module will be aliased as `Vintagenet`
  """

  @doc """
  Indirection for `VintageNet.get/1` (no default)
  """
  @callback get_properties([String.t()]) :: any()

  @doc """
  Give VintageNet a kick to induce reacquiring a (WiFi) connection, by killing
  `VintageNet.RouteManager`
  """
  @callback kick :: :ok

  defmacro __using__(_) do
    mod = implementation(apply(Mix, :env, []), apply(Mix, :target, []))

    quote do
      alias unquote(mod), as: Vintagenet
    end
  end

  def implementation(:test, _), do: VintageHeart.TestingVintagenet
  def implementation(_, :host), do: VintageHeart.StubVintagenet
  def implementation(_, _), do: VintageHeart.RealVintagenet
end


lib/vintage_heart/configuration.ex
defmodule VintageHeart.Configuration do
  @moduledoc """
  Reads optional configuration for Pulse. Except for in during testing
  this library use the configuration app `:vintage_heart` with key, `VintageHeart.Pulse`.

  The following defaults will be used if not otherwise configured:any()

  * `heart_callback?` - `true`. Controls whether the application is registed as the `:heart` callback
  * `poll_interval_millis` - `10_000` - polls every 10 seconds
  * `offline_kick_count` - `24` - After 24, or whatever this is configured to, polls without an internet connection then
  `VintageNet` is kicked causing an attempted reconnection to the `wlan0`.
  * `offline_status_down_count` - `84` - After 84 internet-disconnected polls the status is set to be `:down`. Unless
  `heart_callback?` was false then `:heart` will detect the status and reboot the device.
  * `wizard_hotspot_ip` - {192, 168, 0, 1}. When this is the IP address assume that WiFi is being setup via VintageNetWizard, so connectivity checks are suspended.

  Note that the wizard hotspot id is not validated.
  """

  @default_key VintageHeart.Pulse

  @defaults %{
    heart_callback?: true,
    poll_interval_millis: 10_000,
    offline_kick_count: 24,
    offline_status_down_count: 84,
    wizard_hotspot_ip: {192, 168, 0, 1}
  }

  def heart_callback?(key \\ @default_key), do: val(key, :heart_callback?)
  def poll_interval_millis(key \\ @default_key), do: val(key, :poll_interval_millis)
  def offline_kick_count(key \\ @default_key), do: val(key, :offline_kick_count)
  def offline_status_down_count(key \\ @default_key), do: val(key, :offline_status_down_count)
  def wizard_hotspot_ip(key \\ @default_key), do: val(key, :wizard_hotspot_ip)

  defp val(key, subkey) do
    :vintage_heart
    |> Application.get_env(key, [])
    |> Keyword.get(subkey, @defaults[subkey])
  end
end


lib/vintage_heart/stub_vintagenet.ex
# credo:disable-for-this-file Credo.Check.Readability.LargeNumbers
defmodule VintageHeart.StubVintagenet do
  @moduledoc """
  Pretends to be VintageNet while running on the host machine. Pretends to have a connection
  and kicking does nothing.
  """

  @behaviour VintageHeart.Vintagenet
  @doc """
  """
  @impl true
  def get_properties(properties) do
    case properties do
      ["interface", "wlan0", "addresses"] ->
        [
          %{
            address: {192, 168, 1, 16},
            family: :inet,
            netmask: {255, 255, 255, 0},
            prefix_length: 24,
            scope: :universe
          },
          %{
            address: {65152, 0, 0, 0, 47655, 60415, 65071, 30543},
            family: :inet6,
            netmask: {65535, 65535, 65535, 65535, 0, 0, 0, 0},
            prefix_length: 64,
            scope: :link
          }
        ]

      ["interface", "wlan0", "internet"] ->
        :internet

      _ ->
        nil
    end
  end

  @impl true
  @doc """
  Does nothing
  """
  def kick, do: :ok
end
